### ISSUE TITLE
Epics exported do not have all the links

### ISSUE THREAD

**Issue by kiranramagiri** [solidify/jira-azuredevops-migrator#728]
My config.JSON file has a query that exports all the Epics in the system. Each of the Epic has different work items linked to them as child issues, and the exported JSON file does not have all the child items as links. A few bugs have been exported as links and I don't really see how these are different from the ones not exported. Can someone please help me understand what the issue is.

Config file:

"link-map": {
    "link": [
      {
        "source": "Epic",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Parent",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
 { "source": "Child", 
"target": "System.LinkTypes.Hierarchy-Forward" 
},
{
"source": "relates to",
"target": "System.LinkTypes.Affected By"
},
{
"source": "Duplicate",
"target": "System.LinkTypes.Duplicate-Forward"
},
{
"source": "Dependency",
"target": "System.LinkTypes.Dependency-Forward"
},
{
"source": "Task",
"target": "System.LinkTypes.Hierarchy-Forward"
},
{
"source": "Story",
"target": "System.LinkTypes.Hierarchy-Forward"
},
{
"source": "Bug",
"target": "System.LinkTypes.Hierarchy-Reverse"
},
{
"source": "Dependency",
"target": "System.LinkTypes.Dependency-Reverse"
},

      {
        "source": "Relates",
        "target": "System.LinkTypes.Related-Forward"
      },
{
"source": "Blocks",
"target": "System.LinkTypes.Affected By"
},
{
"source": "is Blocked by",
"target": "System.LinkTypes.Affected By"
},
{
"source": "Clones",
"target": "System.LinkTypes.Related-Forward"
},
{
"source": "is cloned by",
"target": "System.LinkTypes.Related-Reverse"
},
{
"source": "Causes",
"target": "System.LinkTypes.Related-Forward"
},
{
"source": "is caused by",
"target": "System.LinkTypes.Related-Reverse"
},
{
"source": "Predecessor",
"target": "System.LinkTypes.Dependency"
},
      {
        "source": "Duplicate",
        "target": "System.LinkTypes.Duplicate-Forward"
      }
    ]
  },

**Comment by RoyalWulf** [comment]
Good question, I have issues with links as well.  My link-map is different to yours. The link-map in te sample config json file only has a few source/target mappings.  Can we get a full list of what the link-map should look like?

**Comment by Alexander-Hjelm** [comment]
`System.LinkTypes.Related-Forward` and `System.LinkTypes.Related-Reverse` are not valid ADO link types. Use `System.LinkTypes.Related`

See valid link types here: https://learn.microsoft.com/en-us/azure/devops/boards/queries/link-type-reference?view=azure-devops#example

**Comment by kiranramagiri** [comment]
> `System.LinkTypes.Related-Forward` and `System.LinkTypes.Related-Reverse` are not valid ADO link types. Use `System.LinkTypes.Related`
> 
> See valid link types here: https://learn.microsoft.com/en-us/azure/devops/boards/queries/link-type-reference?view=azure-devops#example

I tried following mapping for Blocked, cloned and few other linking's but the links are not exported.

{
"source": "Cloners",
"target": "System.LinkTypes.Duplicate-Forward"
},
{
"source": "duplicates",
"target": "System.LinkTypes.Duplicate-Forward"
},
{
"source": "Dependency",
"target": "System.LinkTypes.Dependency-Forward"
},
{
"source": "Task",
"target": "System.LinkTypes.Hierarchy-Forward"
},
{
"source": "Story",
"target": "System.LinkTypes.Hierarchy-Forward"
},
{
"source": "Bug",
"target": "System.LinkTypes.Hierarchy-Reverse"
},
{
"source": "Dependency",
"target": "System.LinkTypes.Dependency-Reverse"
},

      {
        "source": "Relates",
        "target": "System.LinkTypes.Related-Forward"
      },
{
"source": "Blocks",
"target": "System.LinkTypes.Related-Forward"
},
{
"source": "is blocked by",
"target": "System.LinkTypes.Related-Reverse"
},
{
"source": "clones",
"target": "System.LinkTypes.Related-Forward"
},
{
"source": "is cloned by",
"target": "System.LinkTypes.Related-Reverse"
},
{
"source": "causes",
"target": "System.LinkTypes.Related-Forward"
},
{
"source": "is caused by",
"target": "System.LinkTypes.Related-Reverse"
},
{
"source": "Predecessor",
"target": "System.LinkTypes.Dependency"
},
      {
        "source": "is duplicated by",
        "target": "System.LinkTypes.Related-Reverse"
      }

**Comment by Alexander-Hjelm** [comment]
It looks like you still have some invalid link types, see my previous comment.

**Comment by kiranramagiri** [comment]
Does not tool skip the link type if it's invalid?

**Comment by Alexander-Hjelm** [comment]
Nah, but that could be a good idea for a feature request. I'll create a separate issue for that.

**Comment by kiranramagiri** [comment]
> Nah, but that could be a good idea for a feature request. I'll create a separate issue for that.
You suggest removing all link types barring Parent, Epic, Child, Related and Duplicate? I did not find the mapping for Blocks, Causes, clones and a few other link types. Should I remove all of them?

**Comment by Alexander-Hjelm** [comment]
I suggested that you remove any invalid link types in your config. `System.LinkTypes.Related-Forward` for example
